internal/services/v1/experimental.go

```diff
@@ -202,7 +211,194 @@ func (es *experimentalServer) BulkLoadRelationships(stream v1.ExperimentalServic
 		DispatchCount: 1,
 	})
 
-	return stream.SendAndClose(&v1.BulkLoadRelationshipsResponse{
+	return stream.SendAndClose(&v1.BulkImportRelationshipsResponse{
 		NumLoaded: numWritten,
 	})
 }
+
+func (es *experimentalServer) BulkExportRelationships(
+	req *v1.BulkExportRelationshipsRequest,
+	resp v1.ExperimentalService_BulkExportRelationshipsServer,
+) error {
+	ctx := resp.Context()
+	ds := datastoremw.MustFromContext(ctx)
+
+	var atRevision datastore.Revision
+	var cur dsoptions.Cursor
+
+	if req.OptionalCursor != nil {
+		var err error
+		atRevision, cur, err = decodeCursor(ds, req.OptionalCursor)
+		if err != nil {
+			return shared.RewriteError(ctx, err)
+		}
+	} else {
+		var err error
+		atRevision, _, err = consistency.RevisionFromContext(ctx)
+		if err != nil {
+			return shared.RewriteError(ctx, err)
+		}
+	}
+
+	reader := ds.SnapshotReader(atRevision)
+
+	namespaces, err := reader.ListAllNamespaces(ctx)
+	if err != nil {
+		return shared.RewriteError(ctx, err)
+	}
+
+	// Make sure the namespaces are always in a stable order
+	slices.SortFunc(namespaces, func(
+		lhs datastore.RevisionedDefinition[*core.NamespaceDefinition],
+		rhs datastore.RevisionedDefinition[*core.NamespaceDefinition],
+	) bool {
+		return lhs.Definition.Name < rhs.Definition.Name
+	})
+
+	// Skip the namespaces that are already fully returned
+	for cur != nil && len(namespaces) > 0 && namespaces[0].Definition.Name < cur.ResourceAndRelation.Namespace {
+		namespaces = namespaces[1:]
+	}
+
+	limit := es.defaultBatchSize
+	if req.OptionalLimit > 0 {
+		limit = uint64(req.OptionalLimit)
+	}
+
+	if limit > es.maxBatchSize {
+		limit = es.maxBatchSize
+	}
+
+	relsArray := make([]v1.Relationship, limit)
+	objArray := make([]v1.ObjectReference, limit)
+	subArray := make([]v1.SubjectReference, limit)
+	subObjArray := make([]v1.ObjectReference, limit)
+	caveatArray := make([]v1.ContextualizedCaveat, limit)
+	for i := range relsArray {
+		relsArray[i].Resource = &objArray[i]
+		relsArray[i].Subject = &subArray[i]
+		relsArray[i].Subject.Object = &subObjArray[i]
+	}
+
+	emptyRels := make([]*v1.Relationship, limit)
+
+	for _, ns := range namespaces {
+		rels := emptyRels
+
+		for uint64(len(rels)) == limit {
```

Comment:
Can you comment on exactly how this is functioning? Its non-obvious until you get down to 302